xarray/core/dataset.py

```diff
@@ -3395,9 +3401,16 @@ def roll(self, **shifts):
 
         variables = OrderedDict()
         for name, var in iteritems(self.variables):
-            var_shifts = dict((k, v) for k, v in shifts.items()
-                              if k in var.dims)
-            variables[name] = var.roll(**var_shifts)
+            if name in self.data_vars or roll_coords:
```

Comment:
I think this logic should be updated.
All the variables in `data_vars` should be rolled, independent from `roll_coords` value.
Other variables (variables in `self.coord_names`) may depend on whether `roll_coords` is True or not.